components/core/src/server/oprtenantnode.ts

```diff
@@ -0,0 +1,540 @@
+      listProductsEndpointURL: this.pathToUrl(this.listProductsPath),
+      acceptProductsEndpointURL: this.pathToUrl(this.acceptProductPath),
+      reserveProductsEndpointURL: this.pathToUrl(this.reserveProductPath),
+      rejectProductsEndpointURL: this.pathToUrl(this.rejectProductPath),
+      acceptHistoryEndpointURL: this.pathToUrl(this.historyPath),
+      jwksURL: this.getJwksUrl(),
+      scopesSupported: true,
+    };
+    this.logger =
+      config.logger ?? loglevel.getLogger(`OprHost ${this.hostOrgUrl}`);
+    this.strictCorrectnessChecks = config.strictCorrectnessChecks ?? false;
+    this.integrationApi = new IntegrationApiImpl({
+      host: this,
+      hostOrgUrl: this.hostOrgUrl,
+      model: this.offerModel,
+      storage: this.storage,
+      clock: this.clock,
+      netClient: this.networkClient,
+    });
+    this.serveOrgFile();
+    this.maybeServeJwks();
+    this.serveApiEndpoints();
+  }
+
+  private getPathFromModuleName(moduleName: string) {
+    return moduleName.toLowerCase().replace(/[./]/g, '').replace(/#/g, '/');
+  }
+
+  async start(): Promise<void> {
+    const promises = [];
+    for (const installer of this.integrationInstallers) {
+      const path =
+        installer.mountPath ??
+        'integrations/' +
+          this.getPathFromModuleName(installer.factoryNameSource!);
+      const api = this.integrationApi.namespacedClone(path);
+      promises.push(installer.install(api));
+    }
+    await Promise.all(promises);
+    this.isStartedInternal = true;
+  }
+
+  async destroy(): Promise<void> {
+    await Promise.all(
+      this.integrationInstallers.map(async f =>
+        f.uninstall ? await f.uninstall(this.integrationApi) : undefined
+      )
+    );
+    this.integrationApi.destroy();
+  }
+
+  installOfferProducer(offerProducer: OfferProducer): void {
+    this.offerProducers.push(offerProducer);
+  }
+
+  installFeedConfig(feedConfig: FeedConfig): void {
+    this.feedConfigs.push(feedConfig);
+  }
+
+  private async getFeedProducers(): Promise<Array<OfferProducer>> {
+    return this.feedConfigs.map(feedConfig =>
+      this.getOfferProducerFromFeedConfig(feedConfig)
+    );
+  }
+
+  private getOfferProducerFromFeedConfig(
+    feedConfig: FeedConfig
+  ): OfferProducer {
+    if (!this.networkClient) {
+      throw new StatusError(
+        'An OPR client must be specified to install feeds',
+        'SERVER_CONFIG_ERROR_NO_OPR_CLIENT',
+        500
+      );
+    }
+    return new OprFeedProducer(
+      this.networkClient,
+      feedConfig.organizationUrl,
+      feedConfig.maxUpdateFrequencyMillis
+    );
+  }
+
+  async ingest(): Promise<void> {
+    this.logger.info('Ingesting offers');
+    const producers = [
+      ...(await this.getFeedProducers()),
+      ...this.offerProducers,
+    ];
+    this.logger.info(
+      'Found producers:',
+      producers.map(p => p.id)
+    );
+    for (const producer of producers) {
+      await this.produceOffers(producer);
```

Comment:
Does this run linearly? Could they be parallelized?

Reply:
It could be, but I wasn't sure what the right tradeoff was here. If you parallelize, you risk running out of memory. If you do them in sequence, you risk running out of time. Running out of time is recoverable (if you try again, it'll retry the failed ones), running out of memory is not (if you run out of memory once, you'll probably run out of memory every time), so I decided to play it safe and run them in sequence.